src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -126,49 +134,75 @@ openLevels ::
+        -- When a snapshot is created, merge progress is lost, so we have to
+        -- redo merging work here. NumStepsDone tracks how much merging work
+        -- happened before the snapshot was taken.
+        --
+        -- TODO: this use of supplyMergeCredits is leaky! If a merge completes
+        -- in supplyMergeCredits, then the resulting run is not tracked in the
+        -- registry, and closing the input runs is also not tracked in the
+        -- registry. Note, however, that this bit of code is likely to change in
+        -- #392.
+        forM_ mmmay $ \(NumStepsDone c) -> supplyMergeCredits c incomingRuns
         residentRuns <- V.forM snapResidentRuns $ \rn ->
           allocateTemp reg
             (Run.openFromDisk hfs hbio caching (mkPath rn))
             Run.removeReference
         pure Level{..}
       where
         caching = diskCachePolicyForLevel confDiskCachePolicy ln
+        alloc = bloomFilterAllocForLevel conf ln
 
-        openMergingRun :: SnapMergingRun -> m (MergingRun m h)
-        openMergingRun (SnapMergingRun smrs) = do
-            mrs <- openMergingRunState smrs
-            MergingRun <$> newMVar mrs
+        openMergingRun :: SnapMergingRun -> m (Maybe NumStepsDone, MergingRun m h)
+        openMergingRun (SnapMergingRun mpfl nr smrs) = do
+            (n, mrs) <- openMergingRunState smrs
+            (n,) . MergingRun mpfl nr <$> newMVar mrs
         openMergingRun (SnapSingleRun rn) =
-            SingleRun <$>
+            (Nothing,) . SingleRun <$>
               allocateTemp reg
                 (Run.openFromDisk hfs hbio caching (mkPath rn))
                 Run.removeReference
 
-        openMergingRunState :: SnapMergingRunState -> m (MergingRunState m h)
+        openMergingRunState :: SnapMergingRunState -> m (Maybe NumStepsDone, MergingRunState m h)
         openMergingRunState (SnapCompletedMerge rn) =
-            CompletedMerge <$>
+            (Nothing,) . CompletedMerge <$>
               allocateTemp reg
                 (Run.openFromDisk hfs hbio caching (mkPath rn))
                 Run.removeReference
-        openMergingRunState (SnapOngoingMerge _rns) = do
-            error "openLevels: ongoing merge not yet supported"
+        openMergingRunState (SnapOngoingMerge rns nsd mergeLast) = do
+            rs <- V.forM rns $ \rn ->
+              allocateTemp reg
+                (Run.openFromDisk hfs hbio caching ((mkPath rn)))
+                Run.removeReference
+            nsdVar <- newPrimVar $! unNumStepsDone nsd
+            rn <- uniqueToRunNumber <$> incrUniqCounter uc
+            mergeMaybe <- allocateTemp reg
+              (Merge.new hfs hbio caching alloc mergeLast resolve (mkPath rn) rs)
+              (traverse_ Merge.removeReference)
+            case mergeMaybe of
+              Nothing -> error "openLevels: merges can not be empty"
+              Just m  -> pure (Just nsd, OngoingMerge rs nsdVar m)
```

Comment:
So the merge now has the same number of steps done in `nsdVar` as it had when it was snapshotted (despite at this point not having any merging work done), and then we additionally do that many steps? Doesn't that mean the number of steps is off in the end?

Reply:
Good catch, I'll change that. It will be off I think